# Introduction to Variable Types and Numeric Operators

## Objectives

At the end of this notebook you should be able to:

- differentiate between types of numerics (int, float, complex)
- apply numeric operations
- assign values to variables


## Numeric Variable Types

In this notebook you will repeat the base numeric variable types in Python. There are a number of these base numeric variable types built into Python, and all of them represent a very simple idea, numbers. The ones that we will be looking at today are `ints`, short for integers, `floats`, short for floating point decimal numbers, and `complex`, which contain real and imaginary parts stored as floats.

In [ ]:
14 # Int

In [ ]:
32.2 # Float

In [3]:
complex(5, 6) # complex

(5+6j)

### Types

An important characteristic of Python is that it is a duck typed language. What does this mean? The name duck comes from the classic "If it walks like a duck, and quacks like a duck, then it must be a duck" adage. As applied to our situation, it simply means that Python will determine what it thinks is the best type to call a variable when you use it, unless explicitly told otherwise.

To inspect what type Python thinks a numeric (or anything else is), you can pass it to the `type()` function. Let's see what we get out when we pass numbers of various types to this function.

In [ ]:
type(14)

In [ ]:
type(82.6)

In [ ]:
type(complex(5, 6))

As you can see, Python assumes that a number with no decimal point is an `int`, those with a decimal point a `float`, and (surprise!) those from the `complex()` constructor as `complex`.

Frequently, these subtle differences wont matter too much. However, there will be times when this implementation detail will make you think that something will work, when really it won't. Knowing how to check the type of something will help you solve any of these potential problems.

**Type Questions**

What are the types of the following numerics?
1. `10`
2. `100.32`
3. `complex(-5, -10.2)`

## Numeric Operations

At its base level, Python is really just an awesome calculator that can do way more stuff than addition and subtraction. But, let's focus on that functionality for now.

All of the simple operations that you think should be available are available. Addition, subtraction, multiplication, division and exponentiation are all accessible via `+`, `-`, `*`, `/` and `**`, respectively.

In [ ]:
32 + 23

In [ ]:
19 - 21

In [ ]:
11 * 11

In [5]:
12 / 32

0.375

In [6]:
3 ** 2

9

Perfekt. All of these operations output exactly what we think they would.

Besides those simple operators you can also find on any calculator Python offers two more arithmetic operators which might be new to you: `//` and `%`.       
The double slash `//` is called floor division. All it does is perform division and truncate the result. So where `12 / 32` gave us `0.375`, `12 // 32` cuts off after the `0.` giving us `0`.        

In [18]:
12 // 32 # floor division

0

The last operation that we will go over is the modular division operator, `%` (also called modulo). This operation is the sibling to `//`. As you can see in the following example the `//` gives us the integer number of times that 9 goes into 91. But, there is still a remainder of one. The way we get the remainder of integer division is with the `%` operator.

In [16]:
91 // 9

10

In [17]:
91 % 9  

1

Last but not least, note that you can change the order in which operations get performed by using parenthesis `()`, just as you can in algebra.

In [23]:
46 + 32 * 5

206

In [24]:
(46 + 32) * 5

390

At this point we also want to introduce some simple functions which you will encounter during your daily work as a Data Scientist. (Don't worry if you don't know what a function is. We will cover this in another notebook.)

abs(), min(), max() and round() are useful little functions which will make your life a lot easier. As their names already imply abs() will return the absolut value of a number, min() and max() will return the minimum respectively maximum of a bunch of numbers and round() will round a number to a given amount of decimals.

In [30]:
abs(2.14)

2.14

In [31]:
abs(-2.14)

2.14

In [ ]:
min(14,21,32,11,44,23)

In [ ]:
max(14,21,32,11,44,23)

In [32]:
round(3.14159265359, 3)

3.142

**Operation Questions**

What do you think the results of the following computations will be?
1. `19.0 - 11`
2. `12 * 0.25`
3. `12 ** 2`
4. `25 // 6`
5. `(12 + 8) / 7`
6. `(21 / 3) % 2`

## Variables

One of the most powerful constructs in programming is the ability to store arbitrary values in what we call variables. You can think of variable assignment as giving a name to something so that it can be accessed later by different parts of your program.

In Python, variable assignment occurs with the `=` operator. To assign a value to a variable name (i.e. declare it), you simply put the variable name on the left side of the `=` and the value you want to associate with that variable name on the right side. Once this has happened, you can access the value in the variable simply by using it's name somewhere later in your code or IPython session.

In [ ]:
x = 5
y = 2

In [ ]:
x

In [ ]:
x - 4

In [ ]:
x + y

The name you can give a variable can technically be any contiguous set of characters, but there are some conventions followed in Python and programming in general. Python follows a variable naming convention called snake case. To write something in snake case, simply use a `_` anywhere you would use a space, and make sure every word is lower case. For example, `this_is_a_variable`. Giving variables good names makes your code more readable and therefore maintainable. There is a big difference between seeing a variable called `degrees` and one called `y`. You should strive to give your variables well-defined, succinct names.

There are of course cases where using less than descriptive variable names follows convention and are, therefore, just fine to use. A common example is the use of `i` to keep track of an index (we will cover this idea next week). Because of its prevalent usage for indexing, it is usually easy to understand what is happening in that context when all you see is the varible name `i`. Here, the lack of descriptiveness is okay. The important thing is that the code is **understandable**.

Note that we saw no output from either `x = 5` or `y = 2` above. This is because the return value that would have been printed as output was assigned to the variables `x` and `y`, respectively. This is why we had to view them in the next lines.

A large part of variables' power is the fact that they can change (vary, if you will). This allows us to use a single variable name to keep track of a specific thing throughout the life of a program. Remember how we assigned the value `5` to `x` above? The exact same syntax can be used to change the value stored in the variable.

Say we want to make the value of `x` five more than it currently is. All we need to do is have x be assigned the value that results from adding 5 to `x`.

In [ ]:
x = x + 5
x

Notice how the first line above is formatted. Python knows that the `=` means variable assignment, so when it sees the first line it evaluates the right side of the equals and then puts that value in `x`, even though `x` is part of the calculation on the right side. `x` is now connected with this new value and the old value is gone.

Changing variables in this way occurs so commonly that there is built-in shorthand for it. The result of the first line could have been achieved with `x += 5`. This *syntactic sugar* is available for all the simple operations `+`, `-`, `*`, `/`, `**`, and `%` that we covered earlier.

**Variable Questions**

Consider the following code:

```
x = 82
y = 30
x += y
y = x - 3
x -= y - 5
```

What are the values of x and y after each line?

Person1 is 21 years old, Person2 is 43 years old and Person3 is 19 years old. Define variables for the ages, calculate the mean with them and save the resulte in another variable. Check which type the result has. Can you round the result to 2 decimals?